src/predictor/array_tree_layout.h

```diff
@@ -0,0 +1,273 @@
+        /* We save the node index in the origianl tree to able to continue processing
+         * for nodes not egligable for array layout optimisation. 
+         */
+        nidx_in_tree_[nidx_array - kNodesCount] = nidx;
+    } else {
+      if (IsLeaf(tree, nidx)) {
+        split_index_[nidx_array]  = 0;
+
+        /* 
+         * if the tree is not fully populated, we can reduce transfering costs.
+         * the values for unpopulated part of the tree are set in a way to guarantie
+         * that a moove will always done in "right" direction.
+         * here we exploiting that comparison with nan always results to false.
+         */
+        if constexpr (any_missing) default_left_[nidx_array] = 0;
+        if constexpr (has_categorical) is_cat_[nidx_array] = 0;
+        split_cond_[nidx_array]   = std::numeric_limits<float>::quiet_NaN();
+
+        Populate<depth + 1>(tree, cats, 2 * nidx_array + 2, nidx);
+      } else {
+        if constexpr (any_missing) default_left_[nidx_array] = DefaultLeft(tree, nidx);
+        if constexpr (has_categorical) {
+          is_cat_[nidx_array] = common::IsCat(cats.split_type, nidx);
+          if (is_cat_[nidx_array]) {
+            cat_segment_[nidx_array] = cats.categories.subspan(cats.node_ptr[nidx].beg,
+                                                               cats.node_ptr[nidx].size);
+          }
+        }
+
+        split_index_[nidx_array]  = SplitIndex(tree, nidx);
+        split_cond_[nidx_array]   = SplitCond(tree, nidx);
+
+        /*
+         * LeftChild is used to find if the node is leaf, so it is a valid value,
+         * howerwer RightChild can be invalid in some exotic case.
+         * The tree with invalid right-child can be correctly processed by a classical method,
+         * if the split conditions are propper.
+         * But for array layout invalid RightChild, even unreachable, will lead to memory corruption.
+         * Add check to prevent it.
+         */
+        Populate<depth + 1>(tree, cats, 2 * nidx_array + 1, LeftChild(tree, nidx));
+        bst_node_t right_child = RightChild(tree, nidx);
+        if (right_child != RegTree::kInvalidNodeId) {
+          Populate<depth + 1>(tree, cats, 2 * nidx_array + 2, right_child);
+        }
+      }
+    }
+  }
+
+  bool inline GetDecision(float fvalue, bst_node_t nidx) const {
+    if constexpr (has_categorical) {
+      if (is_cat_[nidx]) {
+       return common::Decision(cat_segment_[nidx], fvalue);
+      } else {
+        return fvalue < split_cond_[nidx];
+      }
+    } else {
+      return fvalue < split_cond_[nidx];
+    }
+  }
+
+ public:
+  /* Ad-hoc value.
+   * Increasing doesn't lead to perf gain, since bottleneck is now at gather instructions.
+   */
+  constexpr static int kMaxNumDeepLevels = 6;
+  static_assert(kNumDeepLevels <= kMaxNumDeepLevels);
+
+  ArrayTreeLayout(const TreeType& tree, RegTree::CategoricalSplitMatrix const &cats) {
+    Populate(tree, cats);
+  }
+
+  /**
+   * @brief
```

Comment:
Also incoherent use of `@brief` and `\param`

Reply:
done